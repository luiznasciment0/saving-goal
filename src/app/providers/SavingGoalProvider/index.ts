import constate from 'constate'
import { useReducer, useCallback } from 'react'

import { MinDate, SavingGoalState, Actions } from 'app/models/SavingGoal'

const currencyMask = (money: number) => {
  return new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: 'USD',
    minimumFractionDigits: 2
  }).format(money)
}

const reducer = (state: SavingGoalState, action: Actions) => {
  switch (action.type) {
    case 'SET_TOTAL_AMOUNT':
      return { ...state, totalAmount: action.money }
    case 'SET_DATE':
      return { ...state, reachDate: action.date }
    case 'SET_MONTHLY_AMOUNT':
      return {
        ...state,
        monthlyAmount: action.amount
      }
  }
}

const useSavingGoal = ({ initialDate }: MinDate) => {
  const [state, dispatch] = useReducer(reducer, {
    reachDate: initialDate,
    totalAmount: '',
    monthlyAmount: ''
  })

  const setAmount = (money: string) => {
    dispatch({ type: 'SET_TOTAL_AMOUNT', money })
  }

  const setDateToReachGoal = (date: Date) => {
    dispatch({ type: 'SET_DATE', date })
  }

  const calcMonthlyAmount = useCallback(
    (dateDiff: number, totalAmount: string) => {
      const initialAmount = totalAmount === '0' ? '$0,00' : `${totalAmount}`

      if (dateDiff === 0) {
        dispatch({ type: 'SET_MONTHLY_AMOUNT', amount: initialAmount })
        return
      }

      const numberTotalAmount = Number(totalAmount.replace(/[,]+/g, ''))

      dispatch({
        type: 'SET_MONTHLY_AMOUNT',
        amount: currencyMask(numberTotalAmount / dateDiff)
      })
    },
    []
  )

  return {
    state,
    effects: {
      setAmount,
      setDateToReachGoal,
      calcMonthlyAmount
    }
  }
}

const [SavingGoalProvider, useStateSavingGoal, useEffectsSavingGoal] = constate(
  useSavingGoal,
  (value) => value.state,
  (value) => value.effects
)

export { SavingGoalProvider, useStateSavingGoal, useEffectsSavingGoal }
