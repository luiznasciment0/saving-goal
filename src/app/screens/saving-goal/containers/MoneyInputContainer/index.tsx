import { useState } from 'react'
import { useRifm } from 'rifm'
import { useEffectsSavingGoal } from 'app/providers/SavingGoalProvider'

import MoneyInput from '../../components/MoneyInput'

const currencyMask = (str: string) => {
  let value = str
  value = value.replace(/\D/g, '')
  value = value.replace(/(\d)(\d{2})$/, '$1.$2')
  value = value.replace(/(?=(\d{3})+(\D))\B/g, ',')

  str = value

  return value
}

const MoneyInputContainer = () => {
  const [value, setValue] = useState('')
  const { setAmount } = useEffectsSavingGoal()

  const setState = (value: string) => {
    setValue(value)
    setAmount(value)
  }

  const handleChange = (value: string) => {
    setState(value)
  }

  const rifm = useRifm({
    value,
    onChange: handleChange,
    format: currencyMask
  })

  return <MoneyInput value={rifm.value} handleChange={rifm.onChange} />
}

export default MoneyInputContainer
