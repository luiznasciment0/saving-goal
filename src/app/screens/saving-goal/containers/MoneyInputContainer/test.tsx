import { render, fireEvent } from '@testing-library/react'
import 'jest-styled-components'

import { SavingGoalProvider } from 'app/providers/SavingGoalProvider'
import MoneyInputContainer from '.'

const setup = () => {
  const utils = render(
    <SavingGoalProvider initialDate={new Date()}>
      <MoneyInputContainer />
    </SavingGoalProvider>
  )
  const input = utils.getByLabelText('Money input')
  return {
    input,
    ...utils
  }
}

describe('<MoneyInput />', () => {
  it('should allow only numbers', () => {
    const { container, input } = setup()

    expect(input).toHaveValue('')
    fireEvent.change(input, { target: { value: 'Test' } })
    expect(input).toHaveValue('')

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should format 1k', () => {
    const { container, input } = setup()

    expect(input).toHaveValue('')
    fireEvent.change(input, { target: { value: '100000' } })
    expect(input).toHaveValue('1,000.00')

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should format 10k', () => {
    const { container, input } = setup()

    expect(input).toHaveValue('')
    fireEvent.change(input, { target: { value: '1000000' } })
    expect(input).toHaveValue('10,000.00')

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should format 100k', () => {
    const { container, input } = setup()

    expect(input).toHaveValue('')
    fireEvent.change(input, { target: { value: '10000000' } })
    expect(input).toHaveValue('100,000.00')

    expect(container.firstChild).toMatchSnapshot()
  })
})
