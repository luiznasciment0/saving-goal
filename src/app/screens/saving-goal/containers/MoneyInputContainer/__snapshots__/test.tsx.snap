// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MoneyInput /> should allow only numbers 1`] = `
.c0 {
  font-size: 1.2rem;
  line-height: 1.6rem;
  font-family: 'Work Sans';
  color: #1e2a32;
  font-weight: 400;
}

<h4
  class="c0"
  color="primaryBlack"
>
  Total amount
</h4>
`

exports[`<MoneyInput /> should format 1k 1`] = `
.c0 {
  font-size: 1.2rem;
  line-height: 1.6rem;
  font-family: 'Work Sans';
  color: #1e2a32;
  font-weight: 400;
}

<h4
  class="c0"
  color="primaryBlack"
>
  Total amount
</h4>
`

exports[`<MoneyInput /> should format 10k 1`] = `
.c0 {
  font-size: 1.2rem;
  line-height: 1.6rem;
  font-family: 'Work Sans';
  color: #1e2a32;
  font-weight: 400;
}

<h4
  class="c0"
  color="primaryBlack"
>
  Total amount
</h4>
`

exports[`<MoneyInput /> should format 100k 1`] = `
.c0 {
  font-size: 1.2rem;
  line-height: 1.6rem;
  font-family: 'Work Sans';
  color: #1e2a32;
  font-weight: 400;
}

<h4
  class="c0"
  color="primaryBlack"
>
  Total amount
</h4>
`
